w32023-to-w3up.test.js

```diff
@@ -0,0 +1,269 @@
+    const migrationEvents = []
+
+    let maxHangSize = 0
+    /**
+     * @param {any} v - promise will resolve with this
+     * @param {number} ms - milliseconds to defer
+     */
+    const defer = (v, ms=1000) => new Promise((resolve) => setTimeout(() => resolve(v)))
+
+    let migrateError
+    await Promise.race([
+      /** run migration and abort on errors */
+      async function () {
+        try {
+          for await (const event of migration) {
+            migrationEvents.push(event)
+            if (aborter.signal.aborted) break
+          }
+        } catch (error) {
+          migrateError = error
+          aborter.abort(error)
+        }
+      }(),
+      /**
+       * Monitor to make sure there are never more hanging connections at carFinder
+       * than the concurrency value, otherwise throw.
+       * End when the number of hanging connection equals concurrency.
+       */
+      async function () {
+        while (await defer(true)) {
+          maxHangSize = Math.max(maxHangSize, hangs.size)
+          if (hangs.size > concurrency) {
+            throw new Error(`migration made a number of pending requests to fetchPart that exceeded options.concurrency, which shouldnt happen`)
+          }
+          if (aborter.signal.aborted) return
+          if (hangs.size === concurrency) {
+            await new Promise((resolve) => setImmediate(resolve))
+            break;
+          }
+        }
+      }(),
+    ])
+
+    assert.equal(hangs.size, concurrency, `hangs.size === concurrency`)
+
+    // we dont promise it will be exactly concurrency + 2, but there shouldn't be any reason to fetch more than that
+    assert.ok(uploads.pulledCount <= (concurrency+2), 'didnt pull more than needed to satisfy concurrency')
+
+    for (const { resolve } of hangs) resolve()
+    assert.equal(hangs.size, 0)
+
+    // tick to allow migration to flow
+    await new Promise((resolve) => setImmediate(resolve))
+
+    // stop the migration
+    aborter.abort()
+
+    assert.ok( ! migrateError, 'migrate asyncgenerator didnt throw')
+    
+    let copiedABlock = false
+    for (const event of migrationEvents) {
+      assert.ok(event.upload instanceof W32023Upload, 'event.upload is a W32023Upload')
+      assert.ok(event.parts instanceof Map, 'event.parts is a Map of part cid to migrated part')
+      assert.equal(event.parts.size, event.upload.parts.length, 'migrated upload has one part for each part cid in input upload')
+      assert.ok(event.add.receipt.out.ok)
+      for (const [cid, migratedPart] of event.parts) {
+        assert.deepEqual(migratedPart.upload, event.upload)
+        assert.ok(migratedPart.upload.parts.includes(cid))
+        assert.ok(migratedPart.add.receipt.out.ok)
+        if (migratedPart.copy) {
+          copiedABlock = true
+        }
+      }
+    }
+    assert.equal(migrationEvents.length, concurrency)
+    assert.equal(maxHangSize, concurrency, `maxHangSize === concurrency`)
+    // @todo this should be true
```

Comment:
handle this @todo. I don't think it will require more functional code, just modifying the mock carFinder to sometimes say 'status:upload'.

But this test is already gnarly so maybe I will just do a separate test that doesn't check concurrency and just mocks out the status=upload case.